feat: DIA-868: Component library + Chip Input

nick-skriabin opened the pull request:
Introducing the UI component library with its first citizen - the ChipInput component.

UI library contains:
- UI (React) component written in TypeScript and styled with SCSS
- Assets library -- icons in SVG format
- Components tests -- Cypress
- Storybook with interactive access to the components and/or components composition

This work was initiated by DIA-868 to support so-called "classes input" (see designs in the ticket). The basis for this component is "MultiEmailInput" from LSE's invitation flow. The component was refactored from JS+Stylus into TS+SCSS, and generalized to be universal.

The work on the component library was initiated due to inability to test components that are not the part of the UI yet.

Note: Name "Classes component" comes from the upcoming PW feature, but the general name for the underlying component is "ChipInput".

### PR fulfills these requirements
- [x] Tests for the changes have been added/updated
- [ ] Docs have been added/updated
- [x] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [x] Self-reviewed and ran all changes on a local instance

### What feature flags were used to cover this change?
None

### What alternative approaches were there?
Implement the whole interface and include a bunch of tickets in a single PR.

### This change affects (describe how if yes)
- [ ] Performance
- [ ] Security
- [x] UX - new generalized chip component to support future work
- [x] DX - introduction of the component library must improve DX by avoiding code duplication, scope creep. It also improves code quality as every component must be covered with tests.

### Does this PR introduce a breaking change?
- [ ] Yes, and covered entirely by feature flag(s)
- [ ] Yes, and covered partially by feature flag(s)
- [x] No
- [ ] Not sure (briefly explain the situation below)

### What level of testing was included in the change?
- [ ] e2e (codecept)
- [x] integration (cypress)
- [ ] unit (jest)

### Which logical domain(s) does this change affect?
Monorepo, PW

nick-skriabin committed:
Generalized chip input component

Introduce storybook
Introduce generalized chip input component
Introduce custom validation for the component

nick-skriabin committed:
Setup stories, link components library to the rest of the project

nick-skriabin committed:
Setting up tests (WIP)

nick-skriabin committed:
Setup component tests

nick-skriabin committed:
fix styles

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/7644877654

nick-skriabin committed:
Tests

nick-skriabin committed:
Cleanup

yyassi-heartex commented on web/libs/storybook/src/components/ChipInput.scss:
```diff
@@ -0,0 +1,96 @@
+@import "../assets/styles/base.scss";
+
+$danger_color: #ff0000;
```
might be a prudent to add the danger color to base.scss to make it available to other components as more components get added here

nick-skriabin replied:
replaced with an existing css var

nick-skriabin committed:
Restructure

hlomzik commented on web/libs/storybook/tests/support/e2e.ts:
```diff
@@ -0,0 +1,24 @@
+import { CURRENT_FLAGS } from '../feature-flags';
+import '@humansignal/frontend-test/cypress/support/e2e'
+
+Cypress.on('uncaught:exception', (err, runnable) => {
```
```suggestion
Cypress.on('uncaught:exception', () => {
```

hlomzik commented on web/libs/storybook/tests/support/e2e.ts:
```diff
@@ -0,0 +1,24 @@
+
+beforeEach(() => {
+  cy.on('uncaught:exception', err => {
+    return !err.message.includes('ResizeObserver loop completed with undelivered notifications.');
```
Most probably we have some other errors, I remember @Gondragos added similar ignore procedure for them. And potentially we could have such error, so better to use array of errors to ignore.

nick-skriabin replied:
ty, reorganized the setup here. I also removed whole e2e section as it's the same thing in cypress. so we only have components tests there now

hlomzik commented on web/libs/storybook/.storybook/main.ts:
```diff
@@ -0,0 +1,34 @@
+        nxViteTsPaths(),
+        svgr({
+          esbuildOptions: {},
+          svgrOptions: { plugins: ['@svgr/plugin-jsx'], ref: true },
```
Not sure how does this work exactly, but in all our other configs we also use `removeViewBox: false` for proper work of some icons.

nick-skriabin replied:
just checked, it doesn't remove it, so we're good

hlomzik replied:
ah, yes, NX does this automatically, that's why it was good in monorepo and failed during back-porting!

hlomzik commented on web/libs/storybook/.storybook/preview-head.html:
```diff
@@ -0,0 +1,36 @@
+    --theme-accent: var(--blue-6);
+    --theme-bg: white; /*var(--gray-2)*/
+
+    --pirmary-on-containte: #052861;
```
Definitely typo in "primary" and I'm not sure I understand the last word

hlomzik replied:
and in other places as well

hlomzik replied:
other places fixed but this exact line is still untouched :D

hlomzik commented on web/libs/storybook/README.md:
Would be great to use this README to describe main principles on how to work with storybook here. For example should we really put components directly here, which ones and why?

nick-skriabin replied:
still working on it. there'l be readme + separate doc about the components lib

hlomzik commented on web/libs/storybook/tests/support/e2e.ts:
```diff
@@ -0,0 +1,24 @@
+      },
+    });
+  });
+  cy.log('Feature flags set');
```
we have both cy.log and console.log here. do we need them?

nick-skriabin replied:
removed it for now completely as we don't have component-level feature flags yet

nick-skriabin committed:
Fix initial value set, update tests

nick-skriabin committed:
More tests

nick-skriabin committed:
Cleanup + ignore unwanted errors

nick-skriabin committed:
Merge branch 'develop' into fb-dia-868/chip-input

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/7670440744

nick-skriabin committed:
Fix typo

nick-skriabin committed:
Ignore tmp

nick-skriabin committed:
Remove excess variable

yyassi-heartex commented on web/libs/storybook/src/assets/icons/index.ts:
```diff
@@ -0,0 +1,3 @@
+/// <reference types="vite-plugin-svgr/client" />
+
```
this is a good opportunity to clean things up
```suggestion
```

nick-skriabin replied:
without it for some reason TS complains about types (can't import things with `?react`). gotta figure out a different way. but this is recommended by the svgr themself

yyassi-heartex replied:
interesting, then lets disregard that - thanks for following up!

nick-skriabin committed:
Remove excess styles

nick-skriabin committed:
Fix docs

nick-skriabin committed:
Merge branch 'develop' into fb-dia-868/chip-input

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/7820203546
